### PR TITLE
Kyber: Implementation in wolfSSL

### PR DESCRIPTION
**Pull request by SparkiDev**
# Description

Put Kyber implementation into wolfSSL.

# Testing

Regression tested kyber.
--enable-kyber no defaults to wolfSSL when liboqs is not specified.

# Checklist

 - [ ] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

### CODE REVIEW

**Comment by SparkiDev** [comment]
retest this please

File: wolfcrypt/src/wc_kyber_poly.c [hunk 1]
```diff
@@ -19,8 +19,3021 @@
...
+        kyber_basemul_mont_add(r, a + i * KYBER_N, b + i * KYBER_N);
+    }
+    kyber_basemul_mont_add(r, a + (kp - 1) * KYBER_N, b + (kp - 1) * KYBER_N);
+}
+
+/******************************************************************************/
+
+/* Initialize Kyber implementation.
+ */
+void kyber_init()
```
Comment (dgarske) [inline wolfcrypt/src/wc_kyber_poly.c:472]:
```wolfcrypt/src/wc_kyber_poly.c:472:16: error: a function declaration without a prototype is deprecated in all versions of C [-Werror,-Wstrict-prototypes]
void kyber_init()
               ^
                void
1 error generated.```

File: wolfcrypt/src/wc_kyber.c [hunk 2]
```diff
@@ -19,8 +19,1249 @@
...
+        ForceZero(key, sizeof(*key));
+    }
+}
+
+/******************************************************************************/
+
+/**
+ * Make a Kyber key object using a random number generator.
+ *
+ * @param  [in, out]  key   Kyber key ovject.
```
Comment (dgarske) [inline wolfcrypt/src/wc_kyber.c:153]: `ovject` -> `object`. 4 places in code.

File: wolfcrypt/src/wc_kyber_poly.c [hunk 3]
```diff
@@ -19,8 +19,3021 @@
...
+    kyber_invntt(v);
+    /* Add errors and message to v and reduce. */
+    for (i = 0; i < KYBER_N; ++i) {
+        sword16 t = v[i] + epp[i] + m[i];
+        v[i] = KYBER_BARRETT_RED(t);
+    }
+}
+
+
+/* Encapsuluate message.
```
Comment (dgarske) [inline wolfcrypt/src/wc_kyber_poly.c:594]: `Encapsuluate` -> `Encapsulate`. Two places

File: wolfcrypt/src/wc_kyber_poly.c [hunk 4]
```diff
@@ -19,8 +19,3021 @@
...
+        p[i + 6] = ETA2_SUB(d, 6);
+        p[i + 7] = ETA2_SUB(d, 7);
+    #endif
+        /* -2 - 1/16, -1 - 4/16, 0 - 6/16, 1 - 4/16, 2 - 1/16  */
+
+        /* Move over used bytes. */
+        r += 4;
+    }
+#else
+    /* Calculate sizteen integer coefficients at a time. */
```
Comment (dgarske) [inline wolfcrypt/src/wc_kyber_poly.c:1395]: sizteen -> sixteen

File: wolfcrypt/src/wc_kyber_poly.c [hunk 5]
```diff
@@ -19,8 +19,3021 @@
...
+    ret = kyber_prf(prf, rand, sizeof(rand), seed);
+    if (ret == 0) {
+        kyber_cbd_eta2_avx2(p, rand);
+    }
+
+    return ret;
+}
+
+/* Get the noise/error by calculating random bytes and sampling to a binomial
+ * deistributions.
```
Comment (dgarske) [inline wolfcrypt/src/wc_kyber_poly.c:1685]: `deistributions` -> `distributions`

**Review by dgarske — CHANGES_REQUESTED** [review]
```
./configure --enable-kyber --enable-smallstack --enable-intelasm && make
...
wolfcrypt/src/wc_kyber_asm.S:12065:21: error: invalid operand for instruction
        movzbl %bl, %r10b
                    ^~~~~
wolfcrypt/src/wc_kyber_asm.S:12066:21: error: invalid operand for instruction
        movzbl %bl, %cl
                    ^~~
wolfcrypt/src/wc_kyber_asm.S:12126:21: error: invalid operand for instruction
        movzbl %bl, %r10b
                    ^~~~~
wolfcrypt/src/wc_kyber_asm.S:12127:21: error: invalid operand for instruction
        movzbl %bl, %cl
                    ^~~
wolfcrypt/src/wc_kyber_asm.S:12187:21: error: invalid operand for instruction
        movzbl %bl, %r10b
                    ^~~~~
wolfcrypt/src/wc_kyber_asm.S:12188:21: error: invalid operand for instruction
        movzbl %bl, %cl
                    ^~~
wolfcrypt/src/wc_kyber_asm.S:12248:21: error: invalid operand for instruction
  CC       examples/client/client-client.o
        movzbl %bl, %r10b
                    ^~~~~
wolfcrypt/src/wc_kyber_asm.S:12249:21: error: invalid operand for instruction
        movzbl %bl, %cl
                    ^~~
wolfcrypt/src/wc_kyber_asm.S:12309:21: error: invalid operand for instruction
        movzbl %bl, %r10b
                    ^~~~~
wolfcrypt/src/wc_kyber_asm.S:12310:21: error: invalid operand for instruction
        movzbl %bl, %cl
                    ^~~
wolfcrypt/src/wc_kyber_asm.S:12370:21: error: invalid operand for instruction
        movzbl %bl, %r10b
                    ^~~~~
wolfcrypt/src/wc_kyber_asm.S:12371:21: error: invalid operand for instruction
        movzbl %bl, %cl
                    ^~~
wolfcrypt/src/wc_kyber_asm.S:12431:21: error: invalid operand for instruction
        movzbl %bl, %r10b
                    ^~~~~
wolfcrypt/src/wc_kyber_asm.S:12432:21: error: invalid operand for instruction
        movzbl %bl, %cl
                    ^~~
wolfcrypt/src/wc_kyber_asm.S:12495:21: error: invalid operand for instruction
        movzbl %bl, %r10b
                    ^~~~~
wolfcrypt/src/wc_kyber_asm.S:12496:21: error: invalid operand for instruction
        movzbl %bl, %cl
                    ^~~
wolfcrypt/src/wc_kyber_asm.S:12686:21: error: invalid operand for instruction
        movzbl %bl, %r10b
                    ^~~~~
wolfcrypt/src/wc_kyber_asm.S:12687:21: error: invalid operand for instruction
        movzbl %bl, %cl
                    ^~~
wolfcrypt/src/wc_kyber_asm.S:12751:21: error: invalid operand for instruction
        movzbl %bl, %r10b
                    ^~~~~
wolfcrypt/src/wc_kyber_asm.S:12752:21: error: invalid operand for instruction
        movzbl %bl, %cl
                    ^~~
wolfcrypt/src/wc_kyber_asm.S:12817:21: error: invalid operand for instruction
        movzbl %bl, %r10b
                    ^~~~~
wolfcrypt/src/wc_kyber_asm.S:12818:21: error: invalid operand for instruction
        movzbl %bl, %cl
                    ^~~
```

```% gcc --version
Apple clang version 15.0.0 (clang-1500.3.9.4)
Target: x86_64-apple-darwin23.3.0
Thread model: posix```

File: wolfcrypt/src/wc_kyber_poly.c (hunk 1 above)
Comment (SparkiDev) [reply wolfcrypt/src/wc_kyber_poly.c:472]: fixed

File: wolfcrypt/src/wc_kyber.c (hunk 2 above)
Comment (SparkiDev) [reply wolfcrypt/src/wc_kyber.c:153]: fixed

File: wolfcrypt/src/wc_kyber_poly.c (hunk 3 above)
Comment (SparkiDev) [reply wolfcrypt/src/wc_kyber_poly.c:594]: fixed

File: wolfcrypt/src/wc_kyber_poly.c (hunk 4 above)
Comment (SparkiDev) [reply wolfcrypt/src/wc_kyber_poly.c:1395]: fixed

**Comment by SparkiDev** [comment]
Assembly code fixed.

File: wolfcrypt/src/wc_kyber_poly.c [hunk 6]
```diff
@@ -19,8 +19,3021 @@
...
+        kyber_cbd_eta2_avx2(poly, rand + 2 * ETA2_RAND_SIZE);
+    }
+
+    return 0;
+}
+#endif
+
+#ifdef WOLFSSL_KYBER1024
+/* Get the noise/error by calculating random bytes and sampling to a binomial
+ * deistributions.
```
Comment (dgarske) [inline wolfcrypt/src/wc_kyber_poly.c:1752]: `deistributions` -> `distributions`. Four places

**Review by dgarske — CHANGES_REQUESTED** [review]
Builds, but fails.

```
./configure --enable-kyber --enable-smallstack --enable-intelasm && make
./wolfcrypt/test/testwolfcrypt
...
KYBER    test failed!
 error L=35378
 [fiducial line numbers: 8247 25532 38178 50526]
```

Many build errors with: `./configure --enable-kyber=small --enable-smallstack --enable-intelasm && make`. Lots also without the intelasm.

```In file included from wolfcrypt/benchmark/benchmark.c:171:
./wolfssl/wolfcrypt/wc_kyber.h:124:18: error: use of undeclared identifier 'KYBER512_K'; did you mean 'KYBER512'?
    sword16 priv[KYBER_MAX_K * KYBER_N];
                 ^~~~~~~~~~~
                 KYBER512
./wolfssl/wolfcrypt/kyber.h:148:37: note: expanded from macro 'KYBER_MAX_K'
#define KYBER_MAX_K                 KYBER512_K
                                    ^
./wolfssl/wolfcrypt/kyber.h:156:5: note: 'KYBER512' declared here
    KYBER512  = 0,
    ^
In file included from wolfcrypt/benchmark/benchmark.c:171:
./wolfssl/wolfcrypt/wc_kyber.h:126:17: error: use of undeclared identifier 'KYBER512_K'; did you mean 'KYBER512'?
    sword16 pub[KYBER_MAX_K * KYBER_N];
                ^~~~~~~~~~~
                KYBER512
./wolfssl/wolfcrypt/kyber.h:148:37: note: expanded from macro 'KYBER_MAX_K'
#define KYBER_MAX_K                 KYBER512_K
                                    ^
./wolfssl/wolfcrypt/kyber.h:156:5: note: 'KYBER512' declared here
    KYBER512  = 0,
    ^
In file included from wolfcrypt/src/wc_kyber_poly.c:31:
./wolfssl/wolfcrypt/wc_kyber.h:124:18: error: use of undeclared identifier 'KYBER512_K'; did you mean 'KYBER512'?
    sword16 priv[KYBER_MAX_K * KYBER_N];
                 ^~~~~~~~~~~
                 KYBER512
./wolfssl/wolfcrypt/kyber.h:148:37: note: expanded from macro 'KYBER_MAX_K'
#define KYBER_MAX_K                 KYBER512_K
                                    ^
./wolfssl/wolfcrypt/kyber.h:156:5: note: 'KYBER512' declared here
    KYBER512  = 0,
    ^
In file included from wolfcrypt/src/wc_kyber_poly.c:31:
./wolfssl/wolfcrypt/wc_kyber.h:126:17: error: use of undeclared identifier 'KYBER512_K'; did you mean 'KYBER512'?
    sword16 pub[KYBER_MAX_K * KYBER_N];
                ^~~~~~~~~~~
                KYBER512
./wolfssl/wolfcrypt/kyber.h:148:37: note: expanded from macro 'KYBER_MAX_K'
#define KYBER_MAX_K                 KYBER512_K
                                    ^
./wolfssl/wolfcrypt/kyber.h:156:5: note: 'KYBER512' declared here
    KYBER512  = 0,
    ^
wolfcrypt/src/wc_kyber_poly.c:1280:35: error: unused parameter 'prf' [-Werror,-Wunused-parameter]
int kyber_gen_matrix(KYBER_PRF_T* prf, sword16* a, int kp, byte* seed,
                                  ^
wolfcrypt/src/wc_kyber_poly.c:1280:49: error: unused parameter 'a' [-Werror,-Wunused-parameter]
int kyber_gen_matrix(KYBER_PRF_T* prf, sword16* a, int kp, byte* seed,
                                                ^
wolfcrypt/src/wc_kyber_poly.c:1280:56: error: unused parameter 'kp' [-Werror,-Wunused-parameter]
int kyber_gen_matrix(KYBER_PRF_T* prf, sword16* a, int kp, byte* seed,
                                                       ^
wolfcrypt/src/wc_kyber_poly.c:1280:66: error: unused parameter 'seed' [-Werror,-Wunused-parameter]
int kyber_gen_matrix(KYBER_PRF_T* prf, sword16* a, int kp, byte* seed,
                                                                 ^
wolfcrypt/src/wc_kyber_poly.c:1281:9: error: unused parameter 'transposed' [-Werror,-Wunused-parameter]
    int transposed)
        ^
wolfcrypt/src/wc_kyber_poly.c:1843:34: error: unused parameter 'prf' [-Werror,-Wunused-parameter]
int kyber_get_noise(KYBER_PRF_T* prf, int kp, sword16* vec1,
                                 ^
wolfcrypt/src/wc_kyber_poly.c:1843:43: error: unused parameter 'kp' [-Werror,-Wunused-parameter]
int kyber_get_noise(KYBER_PRF_T* prf, int kp, sword16* vec1,
                                          ^
wolfcrypt/src/wc_kyber_poly.c:1843:56: error: unused parameter 'vec1' [-Werror,-Wunused-parameter]
int kyber_get_noise(KYBER_PRF_T* prf, int kp, sword16* vec1,
                                                       ^
wolfcrypt/src/wc_kyber_poly.c:1844:14: error: unused parameter 'vec2' [-Werror,-Wunused-parameter]
    sword16* vec2, sword16* poly, byte* seed)
             ^
wolfcrypt/src/wc_kyber_poly.c:1844:29: error: unused parameter 'poly' [-Werror,-Wunused-parameter]
    sword16* vec2, sword16* poly, byte* seed)
                            ^
wolfcrypt/src/wc_kyber_poly.c:1844:41: error: unused parameter 'seed' [-Werror,-Wunused-parameter]
    sword16* vec2, sword16* poly, byte* seed)
                                        ^
13 errors generated.
make[2]: *** [wolfcrypt/src/src_libwolfssl_la-wc_kyber_poly.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
wolfcrypt/benchmark/benchmark.c:9195:13: error: use of undeclared identifier 'KYBER512_CIPHER_TEXT_SIZE'
    byte ct[KYBER_MAX_CIPHER_TEXT_SIZE];
            ^
./wolfssl/wolfcrypt/kyber.h:151:37: note: expanded from macro 'KYBER_MAX_CIPHER_TEXT_SIZE'
#define KYBER_MAX_CIPHER_TEXT_SIZE  KYBER512_CIPHER_TEXT_SIZE
                                    ^
3 errors generated.```

File: wolfcrypt/src/wc_kyber_poly.c (hunk 6 above)
Comment (SparkiDev) [reply wolfcrypt/src/wc_kyber_poly.c]: fixed

**Comment by SparkiDev** [comment]
Can't do --enable-kyber=small.
Must be --enable--kyber=small,all or specific sizes.

**Pull request merged**